tests/Unit/EmogrifierTest.php

```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+     * @param bool $firstSelectorIsMoreSpecific Whether the first selector's rules are expected to prevail
+     * @param string $matchedTagPart Tag (in {@see EmogrifierTest::COMMON_TEST_HTML}) expected to be matched by both
+     *                               selectors, without the closing '>', e.g. '<p class="p-1"'
+     *
+     * @return string[][]
+     */
+    private function getSelectorSpecificityCssDataSet(
+        $firstSelector,
+        $secondSelector,
+        $firstSelectorIsMoreSpecific,
+        $matchedTagPart
+    ) {
+        // The sprintf placeholders %1$s and %2$s will automatically be replaced with CSS declarations
+        // like 'color: red;' or 'color: blue;'.
+        return [
+            $firstSelector . ' { %1$s } ' . $secondSelector . ' { %2$s }',
+            $matchedTagPart . ($firstSelectorIsMoreSpecific ? ' style="%1$s">' : ' style="%2$s">')
+        ];
+    }
+
+    /**
+     * @param string $name Base name for identifying datasets in test results
+     * @param string $firstSelector A selector expression
+     * @param string $secondSelector Some other selector expression
+     * @param string $operator '<', '==' or '>', defining whether the first selector is expected to have, respectively,
+     *                         lower, equal or higher specificity than the second
+     * @param string $matchedTagPart Tag (in {@see EmogrifierTest::COMMON_TEST_HTML}) expected to be matched by both
+     *                               selectors, without the closing '>', e.g. '<p class="p-1"'
+     *
+     * @return string[][]
+     */
+    private function getSelectorSpecificityCssDataSets(
+        $name,
+        $firstSelector,
+        $secondSelector,
+        $operator,
+        $matchedTagPart
+    ) {
+        return [
+            $name . ' (1st selector valid for test)' => $this->getMatchedCssDataSet($firstSelector, $matchedTagPart),
+            $name . ' (2nd selector valid for test)' => $this->getMatchedCssDataSet($secondSelector, $matchedTagPart),
+            // Check first selector matches when its rule is last unless the second has higher specificity
+            $name . ' (1st selector in last rule)' => $this->getSelectorSpecificityCssDataSet(
+                $secondSelector,
+                $firstSelector,
+                $operator === '<',
+                $matchedTagPart
+            ),
+            $name . ' (2nd selector in last rule)' => $this->getSelectorSpecificityCssDataSet(
+                $firstSelector,
+                $secondSelector,
+                $operator === '>',
+                $matchedTagPart
+            )
+        ];
+    }
+
+    /**
+     * @return string[][]
+     */
+    public function selectorSpecificityCssDataProvider()
+    {
+        $selectorData = [
+            ['class more specific than types', '.p-1', 'html body p', '>', '<p class="p-1"'],
+            ['id more specific than class and types', '#p4', 'html body .p-4', '>', '<p class="p-4" id="p4"'],
```

Comment:
id -> ID (in human-readable text, not in code)

Reply:
Changed